### PR TITLE
Updated the CostModel API to take a map of schedule features. 

### PR DESCRIPTION
**Pull request by benoitsteiner**
This change enables implementations of the cost model to customize the way schedule features are handled.

### CODE REVIEW

File: apps/autoscheduler/DefaultCostModel.h [hunk 1]
```diff
@@ -40,9 +40,10 @@ class DefaultCostModel : public CostModel {
                                const MachineParams &params) override;
     void set_pipeline_features(const Runtime::Buffer<float> &, int n);
 
-    // Enqueue a schedule to be evaluated. Returns a buffer of
+    // Enqueue a schedule to be evaluated. The second version of this method returns a buffer of
     // schedule_features that should be filled in by the caller.
-    void enqueue(int ns, Runtime::Buffer<float> *schedule_feats, double *cost_ptr) override;
+    void enqueue(const Internal::Autoscheduler::FunctionDAG &dag, const Halide::Internal::Autoscheduler::StageMapOfScheduleFeatures &schedule_feats, double *cost_ptr) override;
```
Comment (steven-johnson) [inline apps/autoscheduler/DefaultCostModel.h]: see previous comment about really really long lines

**Comment by abadams** [comment]
Tagging @aekul for comment. He was telling me earlier that copying stuff into the schedule features buffer was taking a surprisingly large amount of time, so it would be good to know if this interface affects that one way or the other.

File: apps/autoscheduler/DefaultCostModel.cpp [hunk 2]
```diff
@@ -84,6 +84,50 @@ void DefaultCostModel::set_pipeline_features(const Runtime::Buffer<float> &pipel
     num_cores = n;
 }
 
+void DefaultCostModel::enqueue(const Internal::Autoscheduler::FunctionDAG &dag, const Halide::Internal::Autoscheduler::StageMapOfScheduleFeatures &features, double *cost_ptr) {
```
[the remark this answers is not in the document]
Comment (benoitsteiner) [reply apps/autoscheduler/DefaultCostModel.cpp]: I've updated the code to name these schedule features consistently.

File: apps/autoscheduler/CostModel.h [hunk 3]
```diff
@@ -17,9 +24,9 @@ class CostModel {
     virtual void set_pipeline_features(const Internal::Autoscheduler::FunctionDAG &dag,
                                        const MachineParams &params) = 0;
 
-    // Enqueue a schedule to be evaluated. Returns a buffer of
-    // schedule_features that should be filled in by the caller.
-    virtual void enqueue(int ns, Halide::Runtime::Buffer<float> *schedule_feats, double *cost_ptr) = 0;
+    // Enqueue a schedule to be evaluated. Will annotate the value located at cost_ptr when the evaluation takes place.
+    // Note that the dag argument should correspond to the dag specified previously when callig set_pipeline_features.
+    virtual void enqueue(const Internal::Autoscheduler::FunctionDAG &dag, const Halide::Internal::Autoscheduler::StageMapOfScheduleFeatures &schedule_feats, double *cost_ptr) = 0;
```
[the remark this answers is not in the document]
Comment (benoitsteiner) [reply apps/autoscheduler/CostModel.h]: I've split the line. The feats abbreviation is used in several places in the autoscheduler code, so I've kept it.

File: apps/autoscheduler/DefaultCostModel.cpp [hunk 4]
```diff
@@ -84,6 +84,50 @@ void DefaultCostModel::set_pipeline_features(const Runtime::Buffer<float> &pipel
...
+        // there's no point adding it on to the total because it's
+        // the same across all states.  An underestimate of the
+        // cost for loading from these unscheduled stages is
+        // already baked into the scheduled stages that consume
+        // them.
+        if (stage >= num_stages) break;
+
+        // Load up the schedule features for all stages of this Func.
+        for (auto it = n.stages.rbegin(); it != n.stages.rend(); it++) {
+            internal_assert(features.contains(&*it)) << n.func.name() << "\\
```
[the remark this answers is not in the document]
Comment (benoitsteiner) [reply apps/autoscheduler/DefaultCostModel.cpp]: It looks like a mistake. My editor probably added one backslash to indicate that the line was wrapping around, so I ended up with two backslashes when copying the code.

File: apps/autoscheduler/CostModel.h [hunk 5]
```diff
@@ -5,10 +5,17 @@
...
 #include "HalideBuffer.h"
+#include "PerfectHashMap.h"
 
 // An abstract base class for a cost model.
 namespace Halide {
 
+namespace Internal {
+namespace Autoscheduler {
+typedef PerfectHashMap<FunctionDAG::Node::Stage, ScheduleFeatures> StageMapOfScheduleFeatures;
+}
```
[the remark this answers is not in the document]
Comment (benoitsteiner) [reply apps/autoscheduler/CostModel.h]: Done

**Comment by benoitsteiner** [comment]
This PR only moves code from autoscheduler.cpp to DefaultCostModel.cpp. For better or for worse, there shouldn't be any impact on performance.

File: apps/autoscheduler/DefaultCostModel.cpp [hunk 6]
```diff
@@ -115,9 +117,8 @@ void DefaultCostModel::enqueue(const Internal::Autoscheduler::FunctionDAG &dag,
 
         // Load up the schedule features for all stages of this Func.
         for (auto it = n.stages.rbegin(); it != n.stages.rend(); it++) {
-            internal_assert(features.contains(&*it)) << n.func.name() << "\\
-n";
-            const auto &feat = features.get(&*it);
+            internal_assert(schedule_feats.contains(&*it)) << n.func.name() << "\\n";
```
Comment (aekul) [inline apps/autoscheduler/DefaultCostModel.cpp]: Remove one of the backslashes.

**Comment by aekul** [comment]
Code is just being moved around so this shouldn't impact the copying stuff I mentioned.

File: apps/autoscheduler/DefaultCostModel.cpp (hunk 6 above)
Comment (benoitsteiner) [reply apps/autoscheduler/DefaultCostModel.cpp]: done

File: apps/autoscheduler/CostModel.h [hunk 7]
```diff
@@ -17,9 +24,9 @@ class CostModel {
     virtual void set_pipeline_features(const Internal::Autoscheduler::FunctionDAG &dag,
                                        const MachineParams &params) = 0;
 
-    // Enqueue a schedule to be evaluated. Returns a buffer of
-    // schedule_features that should be filled in by the caller.
-    virtual void enqueue(int ns, Halide::Runtime::Buffer<float> *schedule_feats, double *cost_ptr) = 0;
+    // Enqueue a schedule to be evaluated. Will annotate the value located at cost_ptr when the evaluation takes place.
+    // Note that the dag argument should correspond to the dag specified previously when callig set_pipeline_features.
```
[the remark this answers is not in the document]
Comment (benoitsteiner) [reply apps/autoscheduler/CostModel.h]: done

**Pull request merged**